Adding approach to Queen-Attack (#2860)
exercism/java: exercises/practice/queen-attack/.approaches/simple-comparison/content.md

# Queen Attack Approach

```java
class QueenAttackCalculator {
    private final Queen queen1;
    private final Queen queen2;
    
    QueenAttackCalculator(Queen queen1, Queen queen2) {
        if (queen1 == null || queen2 == null) {
            throw new IllegalArgumentException("You must supply valid positions for both Queens.");
        }
        if (queen1.getRow() == queen2.getRow() && queen1.getColumn() == queen2.getColumn()) {
            throw new IllegalArgumentException("Queens cannot occupy the same position.");
        }
        this.queen1 = queen1;
        this.queen2 = queen2;
    }

    boolean canQueensAttackOneAnother() {
        int rowDifference = Math.abs(queen1.getRow() - queen2.getRow());
        int columnDifference = Math.abs(queen1.getColumn() - queen2.getColumn());
        return rowDifference == 0 || columnDifference == 0 || rowDifference == columnDifference;
    }
}
```

## Approach Steps

1. **Check if Queens are on the Same Row**:
   - If both queens are on the same row, they can attack each other. This is checked by comparing their row values:

     ```java
     if (queen1.getRow() == queen2.getRow()) {
         return true;

>      }

Reviewer: This code example could be confusing as it doesn't appear in the your code sample above. As mentioned in another comment, I think the explanation for checking the row values is mentioned again, so we could probably remove this one,

         return true;
     }
     ```

3. **Check if Queens are on the Same Diagonal**:
   - If both queens are on the same diagonal, they can attack each other. This is checked by comparing the absolute difference between their row and column values:

     ```java
     if (Math.abs(queen1.getRow() - queen2.getRow()) == Math.abs(queen1.getColumn() - queen2.getColumn())) {
         return true;
     }
     ```

4. **Return the Result**:
   - If any of the above checks return `true`, the queens can attack each other. Otherwise, they cannot.

## Explanation

- **Row Check**: We first check if the queens are in the same row. If they are, they can attack each other.
- **Column Check**: Next, we check if the queens are in the same column. If they are, they can attack each other.
- **Diagonal Check**: Lastly, we check if the queens are positioned on the same diagonal. This is determined by comparing the absolute differences in their row and column positions.
- **Final Decision**: If any of these checks return `true`, the queens can attack each other. If none of these conditions are met, they cannot attack each other.

## Additional Explanation for Code

> 

Reviewer: I feel like this _Additional Explanation for Code_ is all the explanation that is needed for the code. The above sections seem to explain how the checks are performed, which are also covered in the explanation for the `canQueensAttackOneAnother`. We could probably make this the _Approach Steps_ section and remove the _Explanation_ section.

1. **Constructor**:

>    In the constructor of `QueenAttackCalculator`, we check if the queens are positioned at valid places.

Reviewer: Since the constructor parameters appears before the checks in the example, I'd suggest opening this paragraph with the fact that it takes two parameters to match the order. I think this would make it easier to follow.

```suggestion
   The constructor of `QueenAttackCalculator` takes two `Queen` objects and checks if they are positioned at valid places.
```
